pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -96,63 +109,132 @@ func (e *DockerRegistryServiceController) Stop() {
+			net.JoinHostPort(fmt.Sprintf("%s.%s.svc.cluster.local", service.Name, service.Namespace), fmt.Sprintf("%d", service.Spec.Ports[0].Port))}
 	}
 
-	return e.defaultDockerURL
+	return []string{e.defaultDockerURL}
 }
 
 // serviceAdded reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
 // changing all interestedURLs
 func (e *DockerRegistryServiceController) serviceAdded(obj interface{}) {
-	service := obj.(*api.Service)
-	if service.Name != e.registryServiceName {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
 		return
 	}
 
-	if err := e.handleLocationChange(e.getServiceLocation(service)); err != nil {
-		util.HandleError(err)
+	service := obj.(*api.Service)
+	if service.Name == e.registryServiceName {
+		e.markForChange()
 	}
 }
 
 // serviceUpdated reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
 // changing all interestedURLs, if needed
 func (e *DockerRegistryServiceController) serviceUpdated(oldObj interface{}, newObj interface{}) {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
+		return
+	}
+
 	oldService := oldObj.(*api.Service)
 	newService := newObj.(*api.Service)
 	if newService.Name != e.registryServiceName {
 		return
 	}
-	if e.getServiceLocation(oldService) == e.getServiceLocation(newService) {
-		return
-	}
 
-	if err := e.handleLocationChange(e.getServiceLocation(newService)); err != nil {
-		util.HandleError(err)
+	oldServiceKeys := e.getServiceLocations(oldService)
+	newServiceKeys := e.getServiceLocations(newService)
+	if !reflect.DeepEqual(oldServiceKeys, newServiceKeys) {
+		e.markForChange()
 	}
+
 }
 
 // serviceDeleted reacts to the docker-registry deletion by updating all the generated dockercfg secrets
 func (e *DockerRegistryServiceController) serviceDeleted(obj interface{}) {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
+		return
+	}
+
 	service, ok := obj.(*api.Service)
 	if !ok {
 		return
 	}
-	if service.Name != e.registryServiceName {
-		return
+	if service.Name == e.registryServiceName {
+		e.markForChange()
 	}
 
-	if err := e.handleLocationChange(e.defaultDockerURL); err != nil {
-		util.HandleError(err)
-	}
 }
 
 // handleLocationChange goes through all service account dockercfg secrets and updates them to point at a new docker-registry location
-func (e *DockerRegistryServiceController) handleLocationChange(serviceLocation string) error {
-	e.dockercfgController.SetDockerURL(serviceLocation)
+func (e *DockerRegistryServiceController) handleLocationChange() error {
+	service, serviceExists, err := e.serviceCache.Get(&api.Service{ObjectMeta: api.ObjectMeta{Namespace: e.registryNamespace, Name: e.registryServiceName}})
```

Comment:
The user is still broken, because everything in etcd for images is broken.
You're not supposed to be changing things on the fly like this - if you
want to do it, it belongs in an upgrade command.  We're going to have to go
write `oadm migrate RegistryRenamed` which updates the image names in the
deployments, and then changes references in RCs, and DCs, and DaemonSets,
and then go fix it.  That is where this code belongs, not in a controller.

On Mon, Jan 18, 2016 at 10:23 AM, David Eads <notifications@github.com>
wrote:

> In pkg/serviceaccounts/controllers/docker_registry_service.go
> <https://github.com/openshift/origin/pull/6385#discussion_r50010046>:
>
> >  }
> >
> >  // handleLocationChange goes through all service account dockercfg secrets and updates them to point at a new docker-registry location
> > -func (e *DockerRegistryServiceController) handleLocationChange(serviceLocation string) error {
> > -	e.dockercfgController.SetDockerURL(serviceLocation)
> > +func (e *DockerRegistryServiceController) handleLocationChange() error {
> > +	service, serviceExists, err := e.serviceCache.Get(&api.Service{ObjectMeta: api.ObjectMeta{Namespace: e.registryNamespace, Name: e.registryServiceName}})
>
> To fix something that should happen ~1-2 times per deployment over its
> entire lifetime. You'll keep processing a large set of changes for
> something that never changes. If an admin wants to change the registry,
> there are lots of other things they have to do (like run a migration on the
> entire etcd database). I want them to run a command that says "hey, go
> ahead and change this" in a controlled fashion, not split that logic into
> multiple locations.
>
> We get emails all the time about, "hey I deleted my registry service and
> it got a new IP and all sorts of insanity resulted". The image secrets
> don't screw them because we have this today. This particular refactor is
> the setup checking routes, which is another case we'd like handle (again,
> because of questions off the list).
>
> The number of changes we're watching is pretty small (one service and
> eventually associated routes), so I don't understand why we wouldn't make
> this automatic.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6385/files#r50010046>.
>